fixit/rules/chained_instance_check.py

```diff
@@ -0,0 +1,188 @@
+            expected_replacement="isinstance(x, (a, b)) or isinstance(y, (c, d)) or isinstance(z, e)",
+        ),
+        Invalid(
+            "isinstance(x, a) or isinstance(x, b) or isinstance(y, c) or isinstance(y, d) "
+            + " or isinstance(z, e) or isinstance(q, f) or isinstance(q, g) or isinstance(q, h)",
+            expected_replacement=(
+                "isinstance(x, (a, b)) or isinstance(y, (c, d)) or isinstance(z, e)"
+                + " or isinstance(q, (f, g, h))"
+            ),
+        ),
+    ]
+
+    def __init__(self, context: CstContext) -> None:
+        super().__init__(context)
+
+        # Since we already unwrap a boolean op's
+        # children
+        self.seen_boolean_operations: Set[cst.BooleanOperation] = set()
+
+    def visit_BooleanOperation(self, node: cst.BooleanOperation) -> None:
+        # Initially match with a partial pattern (in order to ensure
+        # we have enough args to construct more accurate / advanced
+        # pattern).
+
+        if node not in self.seen_boolean_operations and m.matches(
+            node,
+            m.BooleanOperation(right=m.Call(args=[m.AtLeastN(n=1)]), operator=m.Or()),
+        ):
+            calls = self._collect_isinstance_calls(node)
+            if calls is None:
+                return None
+
+            replacement = self._merge_isinstance_calls(calls)
+            if replacement is not None:
+                self.report(node, replacement=replacement)
+
+    def _collect_isinstance_calls(
+        self, node: cst.BooleanOperation
+    ) -> Optional[List[cst.Call]]:
+        expected_call = m.Call(
+            func=m.Name(value="isinstance"),
+            args=[
+                m.Arg(),
+                m.Arg(value=~m.Tuple()),
+            ],
+        )
+        expected_boolop = m.BooleanOperation(operator=m.Or(), right=expected_call)
+
+        seen = []
+        stack: List[cst.Call] = []
+        current = node
+        while m.matches(current, expected_boolop):
+            current = cst.ensure_type(current, cst.BooleanOperation)
+            seen.append(current)
+            stack.insert(0, cst.ensure_type(current.right, cst.Call))
+            current = current.left
+
+        if m.matches(current, expected_call):
+            stack.insert(0, cst.ensure_type(current, cst.Call))
+        else:
+            return None
+
+        self.seen_boolean_operations.update(seen)
+        return stack
+
+    def _merge_isinstance_calls(
+        self, stack: List[cst.Call]
+    ) -> Optional[Union[cst.Call, cst.BooleanOperation]]:
+
+        targets = {}
+        for call in stack:
+            func = cst.ensure_type(call, cst.Call).func
+            for context in self.get_metadata(QualifiedNameProvider, func):
+                if (
+                    context.name != "builtins.isinstance"
+                    or context.source is not QualifiedNameSource.BUILTIN
+                ):
+                    return None
```

Comment:
Why not simply use `self.QualifiedNameProvider.has_name(name="builtins.isinstance", QualifiedNameSource.BUILTIN)` to be simpler?

Reply:
Also, this checks looks redundant since we already check `isinstance` in `_collect_isinstance_calls` and we can simply use `QualifiedNameProvider.has_name(self, name="builtins.isinstance", QualifiedNameSource.BUILTIN)` in `_collect_isinstance_calls`.

Reply:
AFAIK `has_name` check for any appearance, but we want to ensure that it is always the `isinstance` we found (just like `all` instead of `any`). I don't know though if it is the right way or not, can change?

Reply:
Yes, `has_name` checks if any of the qualified name matches and that's how we usually checks the qualified names. https://libcst.readthedocs.io/en/latest/metadata.html#libcst.metadata.QualifiedNameProvider

The `QualifiedName` analysis is based on scope analysis to find assignments from parent scopes and it returns all possible assignments. Sometimes when the code has branches with redefinition of the same name, it could return multiple qualified names. Use `all` can miss lint suggestion on those cases while use `any` can provide the correct suggestion based on our experience.
E.g.
https://github.com/Instagram/LibCST/blob/7ca738bf3964ac48ebb36b1b362153c87c29a706/libcst/codemod/commands/rename.py#L262-L267
https://github.com/Instagram/LibCST/blob/c023fa7c4caff3fd2b3946080f9a58b539b10363/libcst/codemod/commands/convert_namedtuple_to_dataclass.py#L31-L47